File: Kuroba/app/src/main/java/com/github/k1rakishou/chan/ui/controller/AlbumViewController.java
/*
 * KurobaEx - *chan browser https://github.com/K1rakishou/Kuroba-Experimental/
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */
package com.github.k1rakishou.chan.ui.controller;

import android.content.Context;
import android.graphics.Color;
import android.graphics.drawable.Drawable;
import android.view.View;
import android.view.ViewGroup;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.core.content.ContextCompat;
import androidx.recyclerview.widget.GridLayoutManager;
import androidx.recyclerview.widget.RecyclerView;
import androidx.recyclerview.widget.StaggeredGridLayoutManager;

import com.github.k1rakishou.ChanSettings;
import com.github.k1rakishou.chan.R;
import com.github.k1rakishou.chan.controller.Controller;
import com.github.k1rakishou.chan.core.di.component.activity.ActivityComponent;
import com.github.k1rakishou.chan.core.helper.ThumbnailLongtapOptionsHelper;
import com.github.k1rakishou.chan.core.manager.GlobalWindowInsetsManager;
import com.github.k1rakishou.chan.core.manager.WindowInsetsListener;
import com.github.k1rakishou.chan.core.navigation.RequiresNoBottomNavBar;
import com.github.k1rakishou.chan.features.settings.screens.AppearanceSettingsScreen;
import com.github.k1rakishou.chan.ui.cell.AlbumViewCell;
import com.github.k1rakishou.chan.ui.controller.navigation.DoubleNavigationController;
import com.github.k1rakishou.chan.ui.controller.navigation.SplitNavigationController;
import com.github.k1rakishou.chan.ui.theme.widget.ColorizableGridRecyclerView;
import com.github.k1rakishou.chan.ui.toolbar.Toolbar;
import com.github.k1rakishou.chan.ui.toolbar.ToolbarMenuItem;
import com.github.k1rakishou.chan.ui.toolbar.ToolbarMenuSubItem;
import com.github.k1rakishou.chan.ui.view.FastScroller;
import com.github.k1rakishou.chan.ui.view.FastScrollerHelper;
import com.github.k1rakishou.chan.ui.view.post_thumbnail.PostImageThumbnailView;
import com.github.k1rakishou.chan.ui.view.post_thumbnail.ThumbnailView;
import com.github.k1rakishou.common.AndroidUtils;
import com.github.k1rakishou.common.KotlinExtensionsKt;
import com.github.k1rakishou.model.data.descriptor.ChanDescriptor;
import com.github.k1rakishou.model.data.filter.ChanFilterMutable;
import com.github.k1rakishou.model.data.post.ChanPostImage;
import com.github.k1rakishou.persist_state.PersistableChanState;

import org.jetbrains.annotations.NotNull;

import java.util.List;

import javax.inject.Inject;

import kotlin.Unit;

import static com.github.k1rakishou.chan.utils.AppModuleAndroidUtils.dp;
import static com.github.k1rakishou.chan.utils.AppModuleAndroidUtils.getQuantityString;
import static com.github.k1rakishou.chan.utils.AppModuleAndroidUtils.inflate;

public class AlbumViewController
        extends Controller
        implements ImageViewerController.ImageViewerCallback,
        ImageViewerController.GoPostCallback,
        RequiresNoBottomNavBar,
        WindowInsetsListener,
        Toolbar.ToolbarHeightUpdatesCallback {
    private static final int DEFAULT_SPAN_WIDTH = dp(120);

    private static final int ACTION_DOWNLOAD = 0;
    private static final int ACTION_TOGGLE_LAYOUT_MODE = 1;
    private static final int ACTION_TOGGLE_IMAGE_DETAILS = 2;

    private ColorizableGridRecyclerView recyclerView;
    private List<ChanPostImage> postImages;
    private int targetIndex = -1;
    private ChanDescriptor chanDescriptor;

    @Nullable
    private FastScroller fastScroller;

    private AlbumAdapter albumAdapter;
    private final ThreadControllerCallbacks threadControllerCallbacks;

    @Inject
    GlobalWindowInsetsManager globalWindowInsetsManager;
    @Inject
    ThumbnailLongtapOptionsHelper thumbnailLongtapOptionsHelper;

    @Override
    protected void injectDependencies(@NotNull ActivityComponent component) {
        component.inject(this);
    }

    public AlbumViewController(Context context, ThreadControllerCallbacks threadControllerCallbacks) {
        super(context);

        this.threadControllerCallbacks = threadControllerCallbacks;
    }

    @Override
    public void onCreate() {
        super.onCreate();

        // View setup
        view = inflate(context, R.layout.controller_album_view);
        recyclerView = view.findViewById(R.id.recycler_view);
        recyclerView.setHasFixedSize(true);

        albumAdapter = new AlbumAdapter();
        recyclerView.setAdapter(albumAdapter);

        updateRecyclerView(false);

        // Navigation
        Drawable downloadDrawable = ContextCompat.getDrawable(context, R.drawable.ic_file_download_white_24dp);
        downloadDrawable.setTint(Color.WHITE);

        Drawable gridDrawable;

        if (PersistableChanState.getAlbumLayoutGridMode().get()) {
            gridDrawable = ContextCompat.getDrawable(context, R.drawable.ic_baseline_view_quilt_24);
        } else {
            gridDrawable = ContextCompat.getDrawable(context, R.drawable.ic_baseline_view_comfy_24);
        }

        gridDrawable.setTint(Color.WHITE);

        navigation
                .buildMenu(context)
                .withItem(ACTION_TOGGLE_LAYOUT_MODE, gridDrawable, this::toggleLayoutModeClicked)
                .withItem(ACTION_DOWNLOAD, downloadDrawable, this::downloadAlbumClicked)
                .withOverflow(navigationController)
                .withCheckableSubItem(
                        ACTION_TOGGLE_IMAGE_DETAILS,
                        R.string.action_album_show_image_details,
                        true,
                        PersistableChanState.showAlbumViewsImageDetails.get(),
                        this::onToggleAlbumViewsImageInfoToggled)
                .build()
                .build();

        fastScroller = FastScrollerHelper.create(
                FastScroller.FastScrollerControllerType.Album,
                recyclerView,
                null,
                0
        );

        requireNavController().requireToolbar().addToolbarHeightUpdatesCallback(this);
        globalWindowInsetsManager.addInsetsUpdatesListener(this);

        onInsetsChanged();
    }

    private void updateRecyclerView(boolean reloading) {
        SpanInfo spanInfo = getSpanCountAndSpanWidth();

        if (PersistableChanState.getAlbumLayoutGridMode().get()) {
            GridLayoutManager gridLayoutManager = new GridLayoutManager(context, spanInfo.spanCount);
            recyclerView.setLayoutManager(gridLayoutManager);
        } else {
            StaggeredGridLayoutManager staggeredGridLayoutManager = new StaggeredGridLayoutManager(
                    spanInfo.spanCount,
                    StaggeredGridLayoutManager.VERTICAL
            );

            recyclerView.setLayoutManager(staggeredGridLayoutManager);
        }

        recyclerView.setSpanWidth(spanInfo.spanWidth);
        recyclerView.setItemAnimator(null);
        recyclerView.scrollToPosition(targetIndex);

        if (reloading) {
            albumAdapter.refresh();
        }
    }

    private SpanInfo getSpanCountAndSpanWidth() {
        int albumSpanCount = ChanSettings.albumSpanCount.get();
        int albumSpanWith = DEFAULT_SPAN_WIDTH;

        int displayWidth = AndroidUtils.getDisplaySize(context).x;

        if (albumSpanCount == 0) {
            albumSpanCount = displayWidth / DEFAULT_SPAN_WIDTH;
        } else {
            albumSpanWith = displayWidth / albumSpanCount;
        }

        albumSpanCount = AppearanceSettingsScreen.clampColumnsCount(albumSpanCount);

        return new SpanInfo(albumSpanCount, albumSpanWith);
    }

    @Override
    public void onDestroy() {
        super.onDestroy();

        requireNavController().requireToolbar().removeToolbarHeightUpdatesCallback(this);
        globalWindowInsetsManager.removeInsetsUpdatesListener(this);

        if (fastScroller != null) {
            fastScroller.onCleanup();
            fastScroller = null;
        }

        recyclerView.swapAdapter(null, true);
    }

    @Override
    public void onToolbarHeightKnown(boolean heightChanged) {
        if (!heightChanged) {
            return;
        }

        onInsetsChanged();
    }

    @Override
    public void onInsetsChanged() {
        int bottomPadding = globalWindowInsetsManager.bottom();

        if (ChanSettings.getCurrentLayoutMode() == ChanSettings.LayoutMode.SPLIT) {
            bottomPadding = 0;
        }

        KotlinExtensionsKt.updatePaddings(
                recyclerView,
                null,
                FastScrollerHelper.FAST_SCROLLER_WIDTH,
                requireNavController().requireToolbar().getToolbarHeight(),
                bottomPadding
        );
    }

    public void setImages(
            ChanDescriptor chanDescriptor,
            List<ChanPostImage> postImages,
            int index,
            String title
    ) {
        this.chanDescriptor = chanDescriptor;
        this.postImages = postImages;

        navigation.title = title;
        navigation.subtitle = getQuantityString(R.plurals.image, postImages.size(), postImages.size());
        targetIndex = index;
    }

    private void onToggleAlbumViewsImageInfoToggled(ToolbarMenuSubItem subItem) {
        PersistableChanState.showAlbumViewsImageDetails.toggle();
        albumAdapter.refresh();
    }

    private void downloadAlbumClicked(ToolbarMenuItem item) {
        AlbumDownloadController albumDownloadController = new AlbumDownloadController(context);
        albumDownloadController.setPostImages(postImages);
        navigationController.pushController(albumDownloadController);
    }

    private void toggleLayoutModeClicked(ToolbarMenuItem item) {
        PersistableChanState.getAlbumLayoutGridMode().toggle();
        updateRecyclerView(true);

        ToolbarMenuItem menuItem = navigation.findItem(ACTION_TOGGLE_LAYOUT_MODE);

        Drawable gridDrawable;

        if (PersistableChanState.getAlbumLayoutGridMode().get()) {
            gridDrawable = ContextCompat.getDrawable(context, R.drawable.ic_baseline_view_quilt_24);
        } else {
            gridDrawable = ContextCompat.getDrawable(context, R.drawable.ic_baseline_view_comfy_24);
        }

        gridDrawable.setTint(Color.WHITE);
        menuItem.setImage(gridDrawable);
    }

    @Nullable
    @Override
    public ThumbnailView getPreviewImageTransitionView(ChanPostImage postImage) {
        ThumbnailView thumbnail = null;
        for (int i = 0; i < recyclerView.getChildCount(); i++) {
            View view = recyclerView.getChildAt(i);
            if (view instanceof AlbumViewCell) {
                AlbumViewCell cell = (AlbumViewCell) view;
                if (postImage == cell.getPostImage()) {
                    thumbnail = cell.getThumbnailView();
                    break;
                }
            }
        }

        return thumbnail;
    }

    @Override
    public void scrollToImage(ChanPostImage postImage) {
        int index = postImages.indexOf(postImage);
        recyclerView.smoothScrollToPosition(index);
    }

    @Override
    public ImageViewerController.ImageViewerCallback goToPost(ChanPostImage postImage) {
        ThreadController threadController = null;

        if (previousSiblingController instanceof ThreadController) {
            //phone mode
            threadController = (ThreadController) previousSiblingController;
        } else if (previousSiblingController instanceof DoubleNavigationController) {
            //slide mode
            DoubleNavigationController doubleNav = (DoubleNavigationController) previousSiblingController;
            if (doubleNav.getRightController() instanceof ThreadController) {
                threadController = (ThreadController) doubleNav.getRightController();
            }
        } else if (previousSiblingController == null) {
            //split nav has no "sibling" to look at, so we go WAY back to find the view thread controller
            SplitNavigationController splitNav =
                    (SplitNavigationController) this.parentController.parentController.presentedByController;
            threadController = (ThreadController) splitNav.rightController.childControllers.get(0);
            threadController.selectPostImage(postImage);
            //clear the popup here because split nav is weirdly laid out in the stack
            splitNav.popController();
            return threadController;
        }

        if (threadController != null) {
            threadController.selectPostImage(postImage);
            navigationController.popController(false);
            return threadController;
        } else {
            return null;
        }
    }

    private void openImage(AlbumItemCellHolder albumItemCellHolder, ChanPostImage postImage) {
        // Just ignore the showImages request when the image is not loaded
        if (albumItemCellHolder.thumbnailView.getBitmap() != null) {
            final ImageViewerNavigationController imageViewer = new ImageViewerNavigationController(context);
            int index = postImages.indexOf(postImage);
            presentController(imageViewer, false);
            imageViewer.showImages(postImages, index, chanDescriptor, this, this);
        }
    }

    private void showImageLongClickOptions(ChanPostImage postImage) {
        thumbnailLongtapOptionsHelper.onThumbnailLongTapped(
                context,
                true,
                postImage,
                controller -> {
                    presentController(controller);
                    return Unit.INSTANCE;
                },
                chanFilterMutable -> {
                    return Unit.INSTANCE;
                }
        );
    }

    private class AlbumAdapter extends RecyclerView.Adapter<AlbumItemCellHolder> {
        public static final int ALBUM_CELL_TYPE = 1;

        public AlbumAdapter() {
            setHasStableIds(true);
        }

        @Override
        public int getItemViewType(int position) {
            return ALBUM_CELL_TYPE;
        }

        @Override
        public AlbumItemCellHolder onCreateViewHolder(ViewGroup parent, int viewType) {
            View view = inflate(parent.getContext(), R.layout.cell_album_view, parent, false);

            return new AlbumItemCellHolder(view);
        }

        @Override
        public void onBindViewHolder(AlbumItemCellHolder holder, int position) {
            ChanPostImage postImage = postImages.get(position);

            if (postImage != null) {
                boolean canUseHighResCells =
                        ColorizableGridRecyclerView.canUseHighResCells(recyclerView.getCurrentSpanCount());
                boolean isStaggeredGridMode =
                        recyclerView.getLayoutManager() instanceof StaggeredGridLayoutManager;

                holder.cell.bindPostImage(
                        postImage,
                        canUseHighResCells,
                        isStaggeredGridMode,
                        PersistableChanState.getShowAlbumViewsImageDetails().get()
                );
            }
        }

        @Override
        public void onViewRecycled(@NonNull AlbumItemCellHolder holder) {
            holder.cell.unbindPostImage();
        }

        @Override
        public int getItemCount() {
            return postImages.size();
        }

        @Override
        public long getItemId(int position) {
            return position;
        }

        public void refresh() {
            notifyDataSetChanged();
        }
    }

    private class AlbumItemCellHolder
            extends RecyclerView.ViewHolder
            implements View.OnClickListener, View.OnLongClickListener {
        private static final String ALBUM_VIEW_CELL_THUMBNAIL_CLICK_TOKEN = "ALBUM_VIEW_CELL_THUMBNAIL_CLICK";
        private static final String ALBUM_VIEW_CELL_THUMBNAIL_LONG_CLICK_TOKEN = "ALBUM_VIEW_CELL_THUMBNAIL_LONG_CLICK";

        private final AlbumViewCell cell;
        private final PostImageThumbnailView thumbnailView;

        public AlbumItemCellHolder(View itemView) {
            super(itemView);

            cell = (AlbumViewCell) itemView;
            thumbnailView = (PostImageThumbnailView) cell.getThumbnailView();

            thumbnailView.setOnImageClickListener(ALBUM_VIEW_CELL_THUMBNAIL_CLICK_TOKEN, this);
            thumbnailView.setOnImageLongClickListener(ALBUM_VIEW_CELL_THUMBNAIL_LONG_CLICK_TOKEN, this);
        }

        @Override
        public void onClick(View v) {
            int adapterPosition = getAdapterPosition();
            ChanPostImage postImage = postImages.get(adapterPosition);
            openImage(this, postImage);
        }

        @Override
        public boolean onLongClick(View v) {
            int adapterPosition = getAdapterPosition();
            ChanPostImage postImage = postImages.get(adapterPosition);
            showImageLongClickOptions(postImage);
            return true;
        }

    }

    private static class SpanInfo {
        public final int spanCount;
        public final int spanWidth;

        public SpanInfo(int spanCount, int spanWidth) {
            this.spanCount = spanCount;
            this.spanWidth = spanWidth;
        }
    }

    interface ThreadControllerCallbacks {
        void openFiltersController(ChanFilterMutable chanFilterMutable);
    }
}

File: Kuroba/app/src/main/java/com/github/k1rakishou/chan/ui/controller/AlbumViewController.kt
/*
 * KurobaEx - *chan browser https://github.com/K1rakishou/Kuroba-Experimental/
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */
package com.github.k1rakishou.chan.ui.controller

import android.content.Context
import android.graphics.Color
import android.view.View
import android.view.View.OnLongClickListener
import android.view.ViewGroup
import androidx.core.content.ContextCompat
import androidx.recyclerview.widget.GridLayoutManager
import androidx.recyclerview.widget.RecyclerView
import androidx.recyclerview.widget.StaggeredGridLayoutManager
import com.github.k1rakishou.ChanSettings
import com.github.k1rakishou.chan.R
import com.github.k1rakishou.chan.controller.Controller
import com.github.k1rakishou.chan.core.di.component.activity.ActivityComponent
import com.github.k1rakishou.chan.core.helper.ThumbnailLongtapOptionsHelper
import com.github.k1rakishou.chan.core.manager.ChanThreadManager
import com.github.k1rakishou.chan.core.manager.CompositeCatalogManager
import com.github.k1rakishou.chan.core.manager.GlobalWindowInsetsManager
import com.github.k1rakishou.chan.core.manager.WindowInsetsListener
import com.github.k1rakishou.chan.core.navigation.RequiresNoBottomNavBar
import com.github.k1rakishou.chan.core.usecase.FilterOutHiddenImagesUseCase
import com.github.k1rakishou.chan.features.media_viewer.MediaViewerActivity
import com.github.k1rakishou.chan.features.media_viewer.MediaViewerOptions
import com.github.k1rakishou.chan.features.media_viewer.helper.MediaViewerGoToImagePostHelper
import com.github.k1rakishou.chan.features.media_viewer.helper.MediaViewerGoToPostHelper
import com.github.k1rakishou.chan.features.media_viewer.helper.MediaViewerOpenThreadHelper
import com.github.k1rakishou.chan.features.media_viewer.helper.MediaViewerScrollerHelper
import com.github.k1rakishou.chan.features.settings.screens.AppearanceSettingsScreen.Companion.clampColumnsCount
import com.github.k1rakishou.chan.ui.cell.AlbumViewCell
import com.github.k1rakishou.chan.ui.theme.widget.ColorizableGridRecyclerView
import com.github.k1rakishou.chan.ui.toolbar.Toolbar.ToolbarHeightUpdatesCallback
import com.github.k1rakishou.chan.ui.toolbar.ToolbarMenuItem
import com.github.k1rakishou.chan.ui.view.FastScroller
import com.github.k1rakishou.chan.ui.view.FastScrollerHelper
import com.github.k1rakishou.chan.ui.view.FixedLinearLayoutManager
import com.github.k1rakishou.chan.utils.AppModuleAndroidUtils
import com.github.k1rakishou.chan.utils.AppModuleAndroidUtils.dp
import com.github.k1rakishou.common.AndroidUtils
import com.github.k1rakishou.common.isNotNullNorBlank
import com.github.k1rakishou.common.mutableListWithCap
import com.github.k1rakishou.common.updatePaddings
import com.github.k1rakishou.model.data.descriptor.ChanDescriptor
import com.github.k1rakishou.model.data.descriptor.PostDescriptor
import com.github.k1rakishou.model.data.filter.ChanFilterMutable
import com.github.k1rakishou.model.data.post.ChanPostImage
import com.github.k1rakishou.model.util.ChanPostUtils
import com.github.k1rakishou.persist_state.PersistableChanState.albumLayoutGridMode
import com.github.k1rakishou.persist_state.PersistableChanState.showAlbumViewsImageDetails
import kotlinx.coroutines.flow.collect
import kotlinx.coroutines.launch
import okhttp3.HttpUrl
import javax.inject.Inject

class AlbumViewController(
  context: Context,
  private val chanDescriptor: ChanDescriptor,
  private val displayingPostDescriptors: List<PostDescriptor>
) : Controller(context), RequiresNoBottomNavBar, WindowInsetsListener, ToolbarHeightUpdatesCallback {
  private lateinit var recyclerView: ColorizableGridRecyclerView

  private var postImages = mutableListOf<ChanPostImage>()
  private var targetIndex = -1

  private var fastScroller: FastScroller? = null
  private var albumAdapter: AlbumAdapter? = null

  private val spanCountAndSpanWidth: SpanInfo
    get() {
      var albumSpanCount = ChanSettings.albumSpanCount.get()
      var albumSpanWith = DEFAULT_SPAN_WIDTH
      val displayWidth = AndroidUtils.getDisplaySize(context).x

      if (albumSpanCount == 0) {
        albumSpanCount = displayWidth / DEFAULT_SPAN_WIDTH
      } else {
        albumSpanWith = displayWidth / albumSpanCount
      }

      albumSpanCount = clampColumnsCount(albumSpanCount)
      return SpanInfo(albumSpanCount, albumSpanWith)
    }

  @Inject
  lateinit var chanThreadManager: ChanThreadManager
  @Inject
  lateinit var globalWindowInsetsManager: GlobalWindowInsetsManager
  @Inject
  lateinit var thumbnailLongtapOptionsHelper: ThumbnailLongtapOptionsHelper
  @Inject
  lateinit var mediaViewerScrollerHelper: MediaViewerScrollerHelper
  @Inject
  lateinit var mediaViewerGoToImagePostHelper: MediaViewerGoToImagePostHelper
  @Inject
  lateinit var mediaViewerGoToPostHelper: MediaViewerGoToPostHelper
  @Inject
  lateinit var mediaViewerOpenThreadHelper: MediaViewerOpenThreadHelper
  @Inject
  lateinit var filterOutHiddenImagesUseCase: FilterOutHiddenImagesUseCase
  @Inject
  lateinit var compositeCatalogManager: CompositeCatalogManager

  override fun injectDependencies(component: ActivityComponent) {
    component.inject(this)
  }

  override fun onCreate() {
    super.onCreate()

    navigation.title = when (chanDescriptor) {
      is ChanDescriptor.CompositeCatalogDescriptor,
      is ChanDescriptor.CatalogDescriptor -> {
        ChanPostUtils.getTitle(null, chanDescriptor)
      }
      is ChanDescriptor.ThreadDescriptor -> {
        ChanPostUtils.getTitle(
          chanThreadManager.getChanThread(chanDescriptor)?.getOriginalPost(),
          chanDescriptor
        )
      }
    }
    navigation.subtitle = AppModuleAndroidUtils.getQuantityString(R.plurals.image, postImages.size, postImages.size)

    // View setup
    view = AppModuleAndroidUtils.inflate(context, R.layout.controller_album_view)
    recyclerView = view.findViewById(R.id.recycler_view)
    recyclerView.setHasFixedSize(true)
    albumAdapter = AlbumAdapter()
    recyclerView.adapter = albumAdapter
    updateRecyclerView(false)

    // Navigation
    val downloadDrawable = ContextCompat.getDrawable(context, R.drawable.ic_file_download_white_24dp)!!
    downloadDrawable.setTint(Color.WHITE)

    val gridDrawable = if (albumLayoutGridMode.get()) {
      ContextCompat.getDrawable(context, R.drawable.ic_baseline_view_quilt_24)!!
    } else {
      ContextCompat.getDrawable(context, R.drawable.ic_baseline_view_comfy_24)!!
    }

    gridDrawable.setTint(Color.WHITE)

    navigation
      .buildMenu(context)
      .withItem(ACTION_TOGGLE_LAYOUT_MODE, gridDrawable) { item -> toggleLayoutModeClicked(item) }
      .withItem(ACTION_DOWNLOAD, downloadDrawable) { item -> downloadAlbumClicked(item) }
      .withOverflow(navigationController)
      .withCheckableSubItem(
        ACTION_TOGGLE_IMAGE_DETAILS,
        R.string.action_album_show_image_details,
        true,
        showAlbumViewsImageDetails.get()
      ) { onToggleAlbumViewsImageInfoToggled() }
      .build()
      .build()

    fastScroller = FastScrollerHelper.create(
      FastScroller.FastScrollerControllerType.Album,
      recyclerView,
      null
    )

    mainScope.launch {
      mediaViewerScrollerHelper.mediaViewerScrollEventsFlow
        .collect { scrollToImageEvent ->
          val descriptor = scrollToImageEvent.chanDescriptor
          if (descriptor != chanDescriptor) {
            return@collect
          }

          val index = postImages.indexOf(scrollToImageEvent.chanPostImage)
          if (index < 0) {
            return@collect
          }

          scrollToInternal(index)
        }
    }

    mainScope.launch {
      mediaViewerGoToImagePostHelper.mediaViewerGoToPostEventsFlow
        .collect { goToPostEvent ->
          val postImage = goToPostEvent.chanPostImage
          val chanDescriptor = goToPostEvent.chanDescriptor

          popFromNavControllerWithAction(chanDescriptor) { threadController ->
            threadController.selectPostImage(postImage)
          }
        }
    }

    mainScope.launch {
      mediaViewerGoToPostHelper.mediaViewerGoToPostEventsFlow
        .collect { postDescriptor ->
          if (postDescriptor.descriptor != chanDescriptor) {
            return@collect
          }

          popFromNavController(postDescriptor.descriptor)
        }
    }

    if (chanDescriptor is ChanDescriptor.CompositeCatalogDescriptor) {
      mainScope.launch {
        val compositeCatalogName = compositeCatalogManager.byCompositeCatalogDescriptor(chanDescriptor)
          ?.name

        if (compositeCatalogName.isNotNullNorBlank()) {
          navigation.title = compositeCatalogName
          requireNavController().requireToolbar().updateTitle(navigation)
        }
      }
    }

    requireNavController().requireToolbar().addToolbarHeightUpdatesCallback(this)
    globalWindowInsetsManager.addInsetsUpdatesListener(this)
    onInsetsChanged()
  }

  private fun scrollToInternal(scrollPosition: Int) {
    val layoutManager = recyclerView.layoutManager

    if (layoutManager is GridLayoutManager) {
      layoutManager.scrollToPositionWithOffset(scrollPosition, 0)
      return
    }

    if (layoutManager is StaggeredGridLayoutManager) {
      layoutManager.scrollToPositionWithOffset(scrollPosition, 0)
      return
    }

    if (layoutManager is FixedLinearLayoutManager) {
      layoutManager.scrollToPositionWithOffset(scrollPosition, 0)
      return
    }

    recyclerView.scrollToPosition(scrollPosition)
  }

  private fun updateRecyclerView(reloading: Boolean) {
    val spanInfo = spanCountAndSpanWidth
    val staggeredGridLayoutManager = StaggeredGridLayoutManager(
      spanInfo.spanCount,
      StaggeredGridLayoutManager.VERTICAL
    )

    recyclerView.layoutManager = staggeredGridLayoutManager
    recyclerView.setSpanWidth(spanInfo.spanWidth)
    recyclerView.itemAnimator = null
    recyclerView.scrollToPosition(targetIndex)

    if (reloading) {
      albumAdapter?.refresh()
    }
  }

  override fun onDestroy() {
    super.onDestroy()
    requireNavController().requireToolbar().removeToolbarHeightUpdatesCallback(this)
    globalWindowInsetsManager.removeInsetsUpdatesListener(this)

    fastScroller?.onCleanup()
    fastScroller = null

    recyclerView.swapAdapter(null, true)
  }

  override fun onToolbarHeightKnown(heightChanged: Boolean) {
    if (!heightChanged) {
      return
    }

    onInsetsChanged()
  }

  override fun onInsetsChanged() {
    val bottomPaddingDp = calculateBottomPaddingForRecyclerInDp(
      globalWindowInsetsManager = globalWindowInsetsManager,
      mainControllerCallbacks = null
    )

    recyclerView.updatePaddings(
      left = null,
      right = FastScrollerHelper.FAST_SCROLLER_WIDTH,
      top = requireNavController().requireToolbar().toolbarHeight,
      bottom = dp(bottomPaddingDp.toFloat())
    )
  }

  fun tryCollectingImages(initialImageUrl: HttpUrl?): Boolean {
    val (images, index) = collectImages(initialImageUrl)

    if (images.isEmpty()) {
      return false
    }

    val input = FilterOutHiddenImagesUseCase.Input(
      images = images,
      index = index,
      isOpeningAlbum = true,
      postDescriptorSelector = { chanPostImage -> chanPostImage.ownerPostDescriptor }
    )

    val output = filterOutHiddenImagesUseCase.filter(input)
    val filteredImages = output.images
    val newIndex = output.index

    if (filteredImages.isEmpty()) {
      return false
    }

    targetIndex = newIndex

    postImages.clear()
    postImages.addAll(filteredImages)

    return true
  }

  private fun collectImages(initialImageUrl: HttpUrl?): Pair<List<ChanPostImage>, Int> {
    var imageIndexToScroll = 0
    var index = 0

    when (chanDescriptor) {
      is ChanDescriptor.CompositeCatalogDescriptor,
      is ChanDescriptor.CatalogDescriptor -> {
        val postImages = mutableListOf<ChanPostImage>()

        displayingPostDescriptors.forEach { displayingPostDescriptor ->
          val chanPost = chanThreadManager.getPost(displayingPostDescriptor)
          if (chanPost == null) {
            return@forEach
          }

          chanPost.iteratePostImages { chanPostImage ->
            postImages += chanPostImage

            if (initialImageUrl != null && chanPostImage.imageUrl == initialImageUrl) {
              imageIndexToScroll = index
            }

            ++index
          }
        }

        return postImages to imageIndexToScroll
      }
      is ChanDescriptor.ThreadDescriptor -> {
        val chanThread = chanThreadManager.getChanThread(chanDescriptor)
        if (chanThread == null) {
          return emptyList<ChanPostImage>() to imageIndexToScroll
        }

        val postImages = mutableListWithCap<ChanPostImage>(chanThread.postsCount)

        chanThread.iteratePostsOrdered { chanPost ->
          chanPost.iteratePostImages { chanPostImage ->
            postImages += chanPostImage

            if (initialImageUrl != null && chanPostImage.imageUrl == initialImageUrl) {
              imageIndexToScroll = index
            }

            ++index
          }
        }

        return postImages to imageIndexToScroll
      }
    }
  }

  private fun onToggleAlbumViewsImageInfoToggled() {
    val nowChecked = showAlbumViewsImageDetails.toggle()

    navigation.findCheckableSubItem(ACTION_TOGGLE_IMAGE_DETAILS)?.let { subItem ->
      subItem.isChecked = nowChecked
    }

    albumAdapter?.refresh()
  }

  private fun downloadAlbumClicked(item: ToolbarMenuItem) {
    val albumDownloadController = AlbumDownloadController(context)
    albumDownloadController.setPostImages(postImages)
    requireNavController().pushController(albumDownloadController)
  }

  private fun toggleLayoutModeClicked(item: ToolbarMenuItem) {
    albumLayoutGridMode.toggle()
    updateRecyclerView(true)
    val menuItem = navigation.findItem(ACTION_TOGGLE_LAYOUT_MODE)

    val gridDrawable = if (albumLayoutGridMode.get()) {
      ContextCompat.getDrawable(context, R.drawable.ic_baseline_view_quilt_24)!!
    } else {
      ContextCompat.getDrawable(context, R.drawable.ic_baseline_view_comfy_24)!!
    }

    gridDrawable.setTint(Color.WHITE)
    menuItem.setImage(gridDrawable)
  }

  private fun openImage(postImage: ChanPostImage) {
    val index = postImages.indexOf(postImage)
    if (index < 0) {
      return
    }

    when (chanDescriptor) {
      is ChanDescriptor.ICatalogDescriptor -> {
        MediaViewerActivity.catalogMedia(
          context = context,
          catalogDescriptor = chanDescriptor,
          initialImageUrl = postImages[index].imageUrl?.toString(),
          transitionThumbnailUrl = postImages[index].getThumbnailUrl()!!.toString(),
          lastTouchCoordinates = globalWindowInsetsManager.lastTouchCoordinates(),
          mediaViewerOptions = MediaViewerOptions(
            mediaViewerOpenedFromAlbum = true
          )
        )
      }
      is ChanDescriptor.ThreadDescriptor -> {
        MediaViewerActivity.threadMedia(
          context = context,
          threadDescriptor = chanDescriptor,
          postDescriptorList = mapPostImagesToPostDescriptors(),
          initialImageUrl = postImages[index].imageUrl?.toString(),
          transitionThumbnailUrl = postImages[index].getThumbnailUrl()!!.toString(),
          lastTouchCoordinates = globalWindowInsetsManager.lastTouchCoordinates(),
          mediaViewerOptions = MediaViewerOptions(
            mediaViewerOpenedFromAlbum = true
          )
        )
      }
    }
  }

  private fun mapPostImagesToPostDescriptors(): List<PostDescriptor> {
    val duplicateSet = mutableSetOf<PostDescriptor>()

    return postImages.mapNotNull { postImage ->
      if (duplicateSet.add(postImage.ownerPostDescriptor)) {
        return@mapNotNull postImage.ownerPostDescriptor
      }

      return@mapNotNull null
    }
  }

  private fun showImageLongClickOptions(postImage: ChanPostImage) {
    thumbnailLongtapOptionsHelper.onThumbnailLongTapped(
      context = context,
      chanDescriptor = chanDescriptor,
      isCurrentlyInAlbum = true,
      postImage = postImage,
      presentControllerFunc = { controller -> presentController(controller) },
      showFiltersControllerFunc = { },
      openThreadFunc = { postDescriptor ->
        popFromNavController(chanDescriptor)
        mediaViewerOpenThreadHelper.tryToOpenThread(postDescriptor)
      },
      goToPostFunc = {
        popFromNavControllerWithAction(chanDescriptor) { threadController ->
          threadController.selectPostImage(postImage)
        }
      }
    )
  }

  private inner class AlbumAdapter : RecyclerView.Adapter<AlbumItemCellHolder>() {
    private val albumCellType = 1

    init {
      setHasStableIds(true)
    }

    override fun getItemViewType(position: Int): Int {
      return albumCellType
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): AlbumItemCellHolder {
      val view = AppModuleAndroidUtils.inflate(parent.context, R.layout.cell_album_view, parent, false)
      return AlbumItemCellHolder(view)
    }

    override fun onBindViewHolder(holder: AlbumItemCellHolder, position: Int) {
      val postImage = postImages.get(position)
      val canUseHighResCells = ColorizableGridRecyclerView.canUseHighResCells(recyclerView.currentSpanCount)
      val isStaggeredGridMode = !albumLayoutGridMode.get()

      holder.cell.bindPostImage(
        chanDescriptor = chanDescriptor,
        postImage = postImage,
        canUseHighResCells = canUseHighResCells,
        isStaggeredGridMode = isStaggeredGridMode,
        showDetails = showAlbumViewsImageDetails.get()
      )
    }

    override fun onViewRecycled(holder: AlbumItemCellHolder) {
      holder.cell.unbindPostImage()
    }

    override fun getItemCount(): Int {
      return postImages.size
    }

    override fun getItemId(position: Int): Long {
      return position.toLong()
    }

    fun refresh() {
      notifyDataSetChanged()
    }
  }

  private inner class AlbumItemCellHolder(itemView: View) : RecyclerView.ViewHolder(itemView), View.OnClickListener, OnLongClickListener {
    private val ALBUM_VIEW_CELL_THUMBNAIL_CLICK_TOKEN = "ALBUM_VIEW_CELL_THUMBNAIL_CLICK"
    private val ALBUM_VIEW_CELL_THUMBNAIL_LONG_CLICK_TOKEN = "ALBUM_VIEW_CELL_THUMBNAIL_LONG_CLICK"

    val cell = itemView as AlbumViewCell
    val thumbnailView = cell.postImageThumbnailView

    init {
      thumbnailView.setImageClickListener(ALBUM_VIEW_CELL_THUMBNAIL_CLICK_TOKEN, this)
      thumbnailView.setImageLongClickListener(ALBUM_VIEW_CELL_THUMBNAIL_LONG_CLICK_TOKEN, this)
    }

    override fun onClick(v: View) {
      val postImage = postImages.getOrNull(adapterPosition)
        ?: return

      openImage(postImage)
    }

    override fun onLongClick(v: View): Boolean {
      val postImage = postImages.getOrNull(adapterPosition)
        ?: return false

      showImageLongClickOptions(postImage)
      return true
    }

  }

  private class SpanInfo(val spanCount: Int, val spanWidth: Int)

  interface ThreadControllerCallbacks {
    fun openFiltersController(chanFilterMutable: ChanFilterMutable)
  }

  companion object {
    private val DEFAULT_SPAN_WIDTH = dp(120f)
    private const val ACTION_DOWNLOAD = 0
    private const val ACTION_TOGGLE_LAYOUT_MODE = 1
    private const val ACTION_TOGGLE_IMAGE_DETAILS = 2
  }
}